rule/package_directory_mismatch.go

```diff
@@ -0,0 +1,142 @@
+package rule
+
+import (
+	"fmt"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/mgechev/revive/lint"
+)
+
+// PackageDirectoryMismatchRule detects when package name doesn't match directory name.
+type PackageDirectoryMismatchRule struct {
+	ignoredDirs *regexp.Regexp
+}
+
+const defaultIgnoredDirs = "testdata"
+
+// Configure the rule to exclude certain directories.
+func (r *PackageDirectoryMismatchRule) Configure(arguments lint.Arguments) error {
+	if len(arguments) < 1 {
+		return r.buildIgnoreRegex([]string{defaultIgnoredDirs})
+	}
+
+	args, ok := arguments[0].(map[string]any)
+	if !ok {
+		return fmt.Errorf("invalid argument type: expected map[string]any, got %T", arguments[0])
+	}
+
+	for k, v := range args {
+		if !isRuleOption(k, "ignore-directories") {
+			return fmt.Errorf("unknown argument %s for %s rule", k, r.Name())
+		}
+		ignoredDirs, ok := v.([]string)
+		if !ok {
+			return fmt.Errorf("invalid value %v for argument %s of rule %s, expected []string got %T", v, k, r.Name(), v)
+		}
+		return r.buildIgnoreRegex(ignoredDirs)
+	}
+
+	return nil
+}
+
+func (r *PackageDirectoryMismatchRule) buildIgnoreRegex(ignoredDirs []string) error {
```

Comment:
This method name is misleading because the method _builds_ the regexp **and** _assigns_ it to `r.ignoredDirs`.

For the sake of clarity, I try to avoid methods with (hidden) side effects. I propose to change the signature to
```suggestion
func (*PackageDirectoryMismatchRule) buildIgnoreRegex(ignoredDirs []string) (*regexp.Regexp, error) {
```
and let the caller assign the result to `r.ignoredDirs`. Like that, all the updates of the rule config take place in a single method: `Configure`